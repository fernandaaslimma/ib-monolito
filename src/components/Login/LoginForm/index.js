import { connect } from "redux-zero/react";

import LoginForm from "./LoginForm";
import loginActions from "../../../actions/login";

export default connect(
  ({ login }) => ({ login }),
  loginActions
)(LoginForm);
