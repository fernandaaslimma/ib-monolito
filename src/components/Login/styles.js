import styled, { css } from "styled-components";
import { rem, remFontSize, media } from "../../styles/tools";
import { darkestBlue, white, grey50, red } from "../../styles/settings";
import { InputWrapper, InputField, Label } from "../common/Input/styles";

export const Container = styled.div`
  display: flex;
  align-items: stretch;
  min-height: 100vh;

  ${media.md(css`
    display: flex;
    justify-content: center;
    align-items: center;
    padding: ${rem(50)};
    min-height: calc(100vh - ${rem(60)});
  `)};
`;

export const Wrapper = styled.div`
  width: 100%;
  background-color: ${white};
  box-shadow: 0 ${rem(20)} ${rem(40)} ${rem(10)} rgba(224, 224, 224, 0.2);
  position: relative;

  ${media.md(css`
    width: ${rem(443)};
    border-radius: ${rem(8)};
  `)};
`;

export const Content = styled.div`
  text-align: center;
  padding: 0 ${rem(18)};
  margin: ${rem(20)} 0;

  ${media.md(css`
    padding: 0 ${rem(45)};
  `)};

  ${InputWrapper} {
    margin-bottom: ${rem(9)};
  }

  ${InputField} {
    height: ${rem(62)};

    :focus ~ ${Label} {
      top: ${rem(20)};
    }
  }
`;

export const FloatingText = styled.p`
  display: none;

  ${media.md(css`
    bottom: ${rem(35)};
    color: ${white};
    display: block;
    font-family: "Roboto Bold", "Roboto Medium", Roboto;
    font-size: ${remFontSize(99)};
    left: ${rem(12)};
    letter-spacing: ${rem(-1.4)};
    line-height: 1;
    position: absolute;
    text-align: left;
    text-transform: uppercase;
    transform-origin: 0 0;
    transform: rotate(-90deg) translateY(-100%) translateX(-50%);
    user-select: none;
  `)};
`;

export const Title = styled.h2`
  font-family: "Roboto Bold", "Roboto Medium", Roboto;
  font-size: ${remFontSize(21)};
  letter-spacing: ${rem(0.1)};
  color: ${darkestBlue};
  margin-bottom: ${rem(25)};

  ${media.md(css`
    margin-bottom: ${rem(29)};
  `)};
`;

export const ButtonWrapper = styled.div`
  margin: ${rem(20)} 0;
`;

export const IconWrapper = styled.div`
  padding-right: ${rem(8.8)};
`;

export const Header = styled.div`
  text-align: center;
  padding: ${rem(27)} ${rem(34)} ${rem(12)} ${rem(44)};
  border-bottom: solid ${rem(0.7)} ${grey50};
  min-height: ${rem(83)};
`;

export const ErrorMessage = styled.p`
  font-family: Lato;
  font-size: ${remFontSize(13)};
  font-style: italic;
  letter-spacing: ${rem(0.1)};
  text-align: center;
  color: ${red};
  height: ${rem(16)};
  margin-bottom: ${rem(9)};

  ${media.md(css`
    margin-bottom: ${rem(16)};
  `)};
`;
