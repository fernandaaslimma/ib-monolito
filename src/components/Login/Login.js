import React, { Fragment, Component } from "react";
import { func, shape, string, bool } from "prop-types";
import Icon from "../common/Icon/";
import { translate, getLanguage } from "../../utils/i18n";
import { isInternetExplorer } from "../../utils/getNavigator";
import { preflightForInternetExplorer } from "../../services/login";
import LoginForm from "./LoginForm";
import getQueryParam from "../../utils/getQueryParam";
import { redirect } from "../../utils/redirect";

import {
  MOBILE_HASH_PREFIX,
  IMPERSONATE_HASH_PREFIX,
  PT_BR,
  EN_US,
  ZH_CN
} from "../../utils/constants";

import {
  Wrapper,
  Header,
  IconWrapper,
  Title,
  FloatingText,
  Content,
  Container
} from "./styles";

function clearCultureLocation(culture) {
  window.location = window.location.href.replace(`?Culture=${culture}`, "");
}

function hasMobileHash(hash) {
  return hash.includes(MOBILE_HASH_PREFIX);
}

function hasImpersonateHash(hash) {
  return hash.includes(IMPERSONATE_HASH_PREFIX);
}

class Login extends Component {
  componentWillMount() {
    /*
      This request is required for Internet Explorer only
      It makes a preflight request without the headers.
    */
    if (isInternetExplorer()) {
      preflightForInternetExplorer();
    }
  }

  componentDidMount() {
    const {
      openToastr,
      updateEmail,
      setLanguage,
      clearLanguageWithoutClearSessionStorage,
      setOpenBankingInfo
    } = this.props;

    const scope = getQueryParam(location, "scope");
    const intent_id = getQueryParam(location, "intent_id");
    const redirect_uri = getQueryParam(location, "redirect_uri");
    const state = getQueryParam(location, "state");

    let openBankingReceivedInfo = {
      scope,
      intent_id,
      redirect_uri,
      state
    };

    scope &&
      intent_id &&
      redirect_uri &&
      setOpenBankingInfo(openBankingReceivedInfo, location);

    const email = getQueryParam(location, "email");
    const validCultures = [PT_BR, EN_US, ZH_CN];

    setTimeout(() => {
      if (email) {
        updateEmail(email);
        openToastr({
          text: translate("YOUR_PASSWORD_HAS_BEEN_SUCCESSFULLY_REGISTERED"),
          isBelow: false,
          isTop: true
        });
      }
    }, 500);

    const culture = getQueryParam(location, "Culture");

    if (
      culture &&
      validCultures.includes(culture) &&
      culture !== getLanguage()
    ) {
      setLanguage(culture);
      clearLanguageWithoutClearSessionStorage();
      clearCultureLocation(culture);
    }

    culture && clearCultureLocation(culture);
  }

  render() {
    const {
      loginSuccess,
      clearLoginError,
      login,
      openModal,
      location,
      handleMobileHash,
      handleImpersonateHash,
      setJustLoggedIn
    } = this.props;

    const mobileHash = hasMobileHash(location.hash);
    const impersonateHash = hasImpersonateHash(location.hash);

    const state = getQueryParam(location, "state", true);
    const code = getQueryParam(location, "code", true);
    const error = getQueryParam(location, "error", true);

    if (loginSuccess) {
      if ((state && code) || (state && error)) {
        setJustLoggedIn(true);
        redirect(`/home/${location.hash}`);
        return null;
      } else {
        setJustLoggedIn(true);
        redirect("/home");
        return null;
      }
    }

    if (login.error && openModal && login.error.status !== 401) {
      openModal({
        title: translate(
          "WE_ARE_SORRY_BUT_WE_WERE_UNABLE_TO_SERVICE_YOUR_REQUEST"
        ),
        icon: "Attention",
        description: [
          translate(
            "THE_SERVICE_IS_TEMPORARILY_UNAVAILABLE_AND_OUR_ENGINEERS_ARE_WORKING_QUICKLY_TO_RESOLVE_THE_ISSUE"
          ),
          translate("TRY_AGAIN_LATER")
        ],
        onClose: () => clearLoginError()
      });
    }

    if (mobileHash && !loginSuccess) {
      handleMobileHash(location.hash.split(MOBILE_HASH_PREFIX).pop()).then(
        () => {
          redirect("/home");
        }
      );
      return "";
    }

    if (impersonateHash && !loginSuccess) {
      handleImpersonateHash(
        location.hash.split(IMPERSONATE_HASH_PREFIX).pop()
      ).then(() => {
        redirect("/home");
      });
      return "";
    }

    return (
      <Fragment>
        <Container>
          <Wrapper data-test="Login">
            <FloatingText>{translate("LOGIN")}</FloatingText>
            <Header>
              <IconWrapper>
                <Icon type="BrandLogo" />
              </IconWrapper>
            </Header>
            <Content>
              <Title>{translate("SIGN_IN_TO_YOUR_ACCOUNT_BELOW")}</Title>
              <LoginForm />
            </Content>
          </Wrapper>
        </Container>
      </Fragment>
    );
  }
}

Login.defaultProps = {
  login: {},
  loginSuccess: false
};

Login.propTypes = {
  login: shape({
    email: string,
    password: string
  }),
  loginSuccess: bool,
  handleMobileHash: func.isRequired,
  handleImpersonateHash: func.isRequired,
  handleUserSubmit: func.isRequired,
  getUserInfo: func,
  handleUserInput: func.isRequired,
  openModal: func.isRequired,
  openToastr: func.isRequired,
  updateEmail: func.isRequired
};

export default Login;
