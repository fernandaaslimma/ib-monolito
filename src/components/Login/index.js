import { connect } from "redux-zero/react";

import Login from "./Login";
import loginActions from "../../actions/login";
import toastrActions from "../../actions/toastr";
import modalActions from "../../actions/modal";
import settingsActions from "../../actions/settings";
import openBankingActions from "../../actions/openBanking";

export default connect(
  ({
    login,
    loginSuccess,
    openBankingReceivedInfo,
    isConsentFlowConfirmation,
    urlRouteParams
  }) => ({
    login,
    loginSuccess,
    openBankingReceivedInfo,
    isConsentFlowConfirmation,
    urlRouteParams
  }),
  (...actionParams) => ({
    ...loginActions(...actionParams),
    ...toastrActions(...actionParams),
    ...modalActions(...actionParams),
    ...settingsActions(...actionParams),
    ...openBankingActions(...actionParams)
  })
)(Login);
