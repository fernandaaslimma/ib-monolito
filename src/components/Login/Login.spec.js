import React from "react";
import { shallow } from "enzyme";
import Login from "./Login";

jest.mock("../../utils/i18n", () => ({
  translate: id => id,
  withI18n: component => component,
  isZhCN: () => false,
  getLanguage: () => "en-US",
  isPtBR: () => false
}));

jest.mock("../../utils/getNavigator");
jest.mock("../../services/login");

const isInternetExplorerMock = require("../../utils/getNavigator")
  .isInternetExplorer;
const preflightForInternetExplorerMock = require("../../services/login")
  .preflightForInternetExplorer;

jest.mock("../../utils/getQueryParam");
const getQueryParamMock = require("../../utils/getQueryParam").default;

const openToastrMock = jest.fn();
const updateEmailMock = jest.fn();

const defaultProps = {
  login: {
    email: "email@email.com",
    password: "123456"
  },
  loginSuccess: false,
  location: {
    hash: ""
  },
  handleUserSubmit: () => {},
  handleUserInput: () => {},
  setLanguage: jest.fn(),
  setOpenBankingInfo: jest.fn(),
  setJustLoggedIn: jest.fn(),
  handleMobileHash: jest.fn(() => Promise.resolve()),
  handleImpersonateHash: jest.fn(() => Promise.resolve()),
  updateEmail: updateEmailMock,
  openToastr: openToastrMock
};

jest.useFakeTimers();

describe("Login component", () => {
  it("should match snapshot with valid props", () => {
    expect(shallow(<Login {...defaultProps} />)).toMatchSnapshot();
  });

  it("should match snapshot with invalid props", () => {
    const props = Object.assign(defaultProps, {
      login: {
        email: "",
        password: ""
      }
    });
    expect(shallow(<Login {...props} />)).toMatchSnapshot();
  });

  it("should match snapshot when loginSuccess", () => {
    const props = Object.assign(defaultProps, {
      login: {
        email: "",
        password: ""
      },
      loginSuccess: true
    });
    expect(shallow(<Login {...props} />)).toMatchSnapshot();
  });

  it("should match snapshot with error message when login fails", () => {
    const props = Object.assign(defaultProps, {
      login: {
        error: true,
        email: "",
        password: ""
      },
      loginSuccess: false
    });
    expect(shallow(<Login {...props} />)).toMatchSnapshot();
  });

  it("should call openModal if login status is greater than 401", () => {
    const props = Object.assign(defaultProps, {
      login: {
        error: {
          status: 500
        }
      },
      openModal: jest.fn()
    });

    shallow(<Login {...props} />);
    expect(props.openModal).toHaveBeenCalled();
  });

  it("should call clearLoginError when onClose modal function is triggered", () => {
    const props = Object.assign(defaultProps, {
      login: {
        error: {
          status: 500
        }
      },
      clearLoginError: jest.fn(),
      openModal(config) {
        config.onClose();
      }
    });

    shallow(<Login {...props} />);
    expect(props.clearLoginError).toHaveBeenCalled();
  });

  it("should not trigger preflight when browser is not IE", () => {
    isInternetExplorerMock.mockReturnValue(false);
    shallow(<Login {...defaultProps} />);
    expect(preflightForInternetExplorerMock).not.toHaveBeenCalled();
  });

  it("should trigger preflight when browser is IE", () => {
    isInternetExplorerMock.mockReturnValue(true);
    shallow(<Login {...defaultProps} />);
    expect(preflightForInternetExplorerMock).toHaveBeenCalled();
  });

  describe("componentDidMount", () => {
    describe("when email exists", () => {
      it("should invoke updateEmail", () => {
        getQueryParamMock.mockReturnValue("email");
        shallow(<Login {...defaultProps} />);

        jest.advanceTimersByTime(500);

        expect(updateEmailMock).toHaveBeenCalledWith("email");
      });

      it("should invoke openToastr", () => {
        getQueryParamMock.mockReturnValue("email");
        shallow(<Login {...defaultProps} />);

        jest.advanceTimersByTime(500);

        expect(openToastrMock).toHaveBeenCalledWith({
          text: "YOUR_PASSWORD_HAS_BEEN_SUCCESSFULLY_REGISTERED",
          isBelow: false,
          isTop: true
        });
      });
    });

    describe("when email doens't exist", () => {
      it("should invoke updateEmail", () => {
        const updateEmailMock = jest.fn();
        getQueryParamMock.mockReturnValue(null);
        shallow(<Login {...defaultProps} />);

        expect(updateEmailMock).not.toHaveBeenCalled();
      });

      it("should invoke openToastr", () => {
        const openToastrMock = jest.fn();
        getQueryParamMock.mockReturnValue(null);
        shallow(<Login {...defaultProps} />);

        expect(openToastrMock).not.toHaveBeenCalled();
      });
    });
  });

  describe("when login from mobile", () => {
    it("should handle mobile hash", () => {
      shallow(<Login {...defaultProps} location={{ hash: "#mobile_12345" }} />);

      expect(defaultProps.handleMobileHash).toHaveBeenCalledWith("12345");
    });
  });

  describe("when login with impersonate hash", () => {
    it("should handle impersonate hash", () => {
      shallow(
        <Login {...defaultProps} location={{ hash: "#impersonate_12345" }} />
      );

      expect(defaultProps.handleImpersonateHash).toHaveBeenCalledWith("12345");
    });
  });
});
