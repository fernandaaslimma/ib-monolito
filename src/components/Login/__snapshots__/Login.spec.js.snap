// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should match snapshot when loginSuccess 1`] = `""`;

exports[`Login component should match snapshot with error message when login fails 1`] = `
<Fragment>
  <styled.div>
    <styled.div
      data-test="Login"
    >
      <styled.p>
        LOGIN
      </styled.p>
      <styled.div>
        <styled.div>
          <Icon
            color={null}
            cursor={false}
            fixCrossAxisAlign={null}
            height={null}
            type="BrandLogo"
            width={null}
          />
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.h2>
          SIGN_IN_TO_YOUR_ACCOUNT_BELOW
        </styled.h2>
        <ConnectWrapper />
      </styled.div>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`Login component should match snapshot with invalid props 1`] = `
<Fragment>
  <styled.div>
    <styled.div
      data-test="Login"
    >
      <styled.p>
        LOGIN
      </styled.p>
      <styled.div>
        <styled.div>
          <Icon
            color={null}
            cursor={false}
            fixCrossAxisAlign={null}
            height={null}
            type="BrandLogo"
            width={null}
          />
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.h2>
          SIGN_IN_TO_YOUR_ACCOUNT_BELOW
        </styled.h2>
        <ConnectWrapper />
      </styled.div>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`Login component should match snapshot with valid props 1`] = `
<Fragment>
  <styled.div>
    <styled.div
      data-test="Login"
    >
      <styled.p>
        LOGIN
      </styled.p>
      <styled.div>
        <styled.div>
          <Icon
            color={null}
            cursor={false}
            fixCrossAxisAlign={null}
            height={null}
            type="BrandLogo"
            width={null}
          />
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.h2>
          SIGN_IN_TO_YOUR_ACCOUNT_BELOW
        </styled.h2>
        <ConnectWrapper />
      </styled.div>
    </styled.div>
  </styled.div>
</Fragment>
`;
